### PR TITLE
[Merged by Bors] - refactor: generalise the Ruzsa triangle inequality to non-abelian groups

### PR DESCRIPTION
**Pull request by YaelDillies**
The proof works just fine, and the other statements too, up to turning some `*` around. Add supporting lemmas of the form `op '' (s * t) = op '' t * op '' s`. Use the new `#` notation for `Finset.card`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean [hunk 1]
```diff
@@ -141,80 +183,80 @@ theorem ruzsa_triangle_inequality_mul_mul_mul (A B C : Finset α) :
...
-  refine (Nat.cast_le.2 <| pluennecke_petridis_inequality_mul _ hAB).trans ?_
-  rw [cast_mul]
-  gcongr
+  refine le_of_mul_le_mul_right ?_ (by positivity : (0 : ℚ≥0) < #A)
+  calc
+    ((#(A * B ^ (n + 1))) * #A : ℚ≥0)
+      = #(A * B * B ^ n) * #A := by rw [_root_.pow_succ', ← mul_assoc]
+    _ ≤ #(A * B) * #(A * B ^ n) := mod_cast pluennecke_petridis_inequality_mul _ hAB
+    _ ≤ #(A * B) * ((#(A * B) / #A) ^ n * #A) := by gcongr
+    _ = (#(A * B) / #A) ^ (n + 1) * #A * #A := by field_simp; ring
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:219]: nice!

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean [hunk 2]
```diff
@@ -141,80 +183,80 @@ theorem ruzsa_triangle_inequality_mul_mul_mul (A B C : Finset α) :
...
-  push_cast
-  rw [mul_comm _ C]
-  refine (mul_le_mul (card_mul_pow_le (mul_aux hC.1 hC.2 hCA) _)
-    (card_mul_pow_le (mul_aux hC.1 hC.2 hCA) _) (zero_le _) (zero_le _)).trans ?_
-  rw [mul_mul_mul_comm, ← pow_add, ← mul_assoc]
-  gcongr ((?_ ^ _) * Nat.cast ?_) * _
-  · exact hCA _ hA'
-  · exact card_le_card hC.2
+  obtain ⟨hC, hCA⟩ := hC
+  refine le_of_mul_le_mul_right ?_ (by positivity : (0 : ℚ≥0) < #C)
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:232]: hmm this makes me think `le_of_mul_le_mul_right` wants an explicit argument

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean [hunk 3]
```diff
@@ -141,80 +183,80 @@ theorem ruzsa_triangle_inequality_mul_mul_mul (A B C : Finset α) :
...
+  calc
+    (#(B ^ m / B ^ n) * #C : ℚ≥0)
+      ≤ #(B ^ m * C) * #(B ^ n * C) := mod_cast ruzsa_triangle_inequality_div_mul_mul ..
+    _ = #(C * B ^ m) * #(C * B ^ n) := by simp_rw [mul_comm]
+    _ ≤ ((#(C * B) / #C) ^ m * #C) * ((#(C * B) / #C : ℚ≥0) ^ n * #C) := by
+      gcongr <;> exact card_mul_pow_le (mul_aux hC hCA hCmin) _
+    _ = (#(C * B) / #C) ^ (m + n) * #C * #C := by ring
+    _ ≤ (#(A * B) / #A) ^ (m + n) * #A * #C := by gcongr ((?_ ^ _) * #?_) * _; exact hCmin _ hA'
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:240]:
```suggestion
    _ ≤ (#(A * B) / #A) ^ (m + n) * #A * #C := by gcongr (?_ ^ _) * #?_ * _; exact hCmin _ hA'
```
I guess?

**Review by b-mehta — APPROVED** [review]
bors d+

File: Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean [hunk 4]
```diff
@@ -581,6 +591,15 @@ theorem subset_mul {s t : Set α} :
 theorem image_mul [DecidableEq β] : (s * t).image (f : α → β) = s.image f * t.image f :=
   image_image₂_distrib <| map_mul f
 
+@[to_additive]
+lemma image_op_mul (s t : Finset α) : (s * t).image op = t.image op * s.image op :=
+  image_image₂_antidistrib op_mul
+
+@[to_additive]
+lemma map_op_mul (s t : Finset α) :
```
Comment (b-mehta) [inline Mathlib/Algebra/Group/Pointwise/Finset/Basic.lean:599]: Oh, let's add Set versions of these too if they don't already exist

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean [hunk 5]
```diff
@@ -50,31 +56,67 @@ theorem ruzsa_triangle_inequality_div_div_div (A B C : Finset α) :
...
-  rw [← div_inv_eq_mul, ← card_inv B, ← card_inv (B * C), mul_inv, ← div_eq_mul_inv]
+theorem ruzsa_triangle_inequality_div_mul_mul (A B C : Finset G) :
+    #(A / C) * #B ≤ #(A * B) * #(C * B) := by
+  rw [← div_inv_eq_mul, ← card_inv B, ← card_inv (C * B), mul_inv_rev, ← div_eq_mul_inv]
   exact ruzsa_triangle_inequality_div_div_div _ _ _
 
+/-- **Ruzsa's triangle inequality**. Invmul-mul-mul version. -/
+@[to_additive "**Ruzsa's triangle inequality**. Negadd-add-add version."]
+theorem ruzsa_triangle_inequality_inv_mul_mul_mul (A B C : Finset G) :
+    #(C⁻¹ * A) * #B ≤ #(B * A) * #(B * C) := by
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:77]: Wait, you want me to move the `#B` factor argument around?

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean [hunk 6]
```diff
@@ -50,31 +56,67 @@ theorem ruzsa_triangle_inequality_div_div_div (A B C : Finset α) :
...
+  simpa [div_eq_mul_inv, ← map_op_mul, ← map_op_inv] using
+    ruzsa_triangle_inequality_div_mul_mul (G := Gᵐᵒᵖ) (A.map opEquiv.toEmbedding)
+      (B.map opEquiv.toEmbedding) (C.map opEquiv.toEmbedding)
+
 /-- **Ruzsa's triangle inequality**. Mul-div-mul version. -/
 @[to_additive "**Ruzsa's triangle inequality**. Add-sub-add version."]
-theorem ruzsa_triangle_inequality_mul_div_mul (A B C : Finset α) :
-    (A * C).card * B.card ≤ (A / B).card * (B * C).card := by
+theorem ruzsa_triangle_inequality_mul_div_mul (A B C : Finset G) :
+    #(A * C) * #B ≤ #(A / B) * #(B * C) := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:85]: What side should the `B` be on here then?

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:232]: Possible, but out of scope ;)

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean (hunk 5 above)
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:77]: I am not sure this is more consistent than before (and you would have to explain https://github.com/leanprover-community/mathlib4/pull/18145#discussion_r1815470237), so I will keep it like that for now

**Comment by YaelDillies** [comment]
bors merge

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean (hunk 6 above)
Comment (b-mehta) [reply Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:85]: This case is ambiguous, so you can put it where you like

File: Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean (hunk 5 above)
[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/Combinatorics/Additive/PluenneckeRuzsa.lean:77]: Yes, exactly, moving the `#B` argument is what I'd like you to move here. This is for consistency with distributivity lemmas: `B` is on the right on the lhs here, but on the left in both parts of the rhs.

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**